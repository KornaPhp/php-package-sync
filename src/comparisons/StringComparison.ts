import { ComparisonKind } from '../types/FileComparisonResult';
import { Comparison } from './Comparison';

const { compareTwoStrings } = require('string-similarity');

export class StringComparison extends Comparison {
    public compare(requiredScore: number | null): Comparison {
        this.score = compareTwoStrings(this.file?.contents ?? '', this.repoFile?.contents ?? '');
        this.comparisonPassed = this.meetsRequirement(requiredScore ?? 0);

        if (!this.passed()) {
            this.createIssue(null);
        }

        return this;
    }

    public getKind(): ComparisonKind {
        return ComparisonKind.FILE_NOT_SIMILAR_ENOUGH;
    }

    public meetsRequirement(percentage) {
        if (percentage === null || percentage === 0) {
            percentage = this.file?.requiredScores?.similar ?? 15;
        }

        return super.meetsRequirement(percentage);
    }
}
