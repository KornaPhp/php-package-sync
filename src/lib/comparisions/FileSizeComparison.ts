/* eslint-disable no-unused-vars */

export class FileSizeComparison {
    static create(filesize1: number, filesize2: number) {
        return new FileSizeComparison(filesize1, filesize2);
    }

    constructor(public filesize1: number, public filesize2: number) {
        //
    }

    get difference(): number {
        return this.filesize1 - this.filesize2;
    }

    get percentage(): number {
        if (this.difference === 0) {
            return 0;
        }

        return (this.difference / this.filesize1) * 100;
    }

    public meetsRequirement(percentage = 15): boolean {
        if (this.difference === 0) {
            return false;
        }

        const pctDecimal = percentage * 0.01;

        return this.difference >= Math.round(this.filesize1 * pctDecimal);
    }

    public format(decimals = 2, suffix = '%'): string {
        return this.percentage.toFixed(decimals) + suffix;
    }
}

export function compareFileSizes(filesize1: number, filesize2: number): FileSizeComparison {
    return FileSizeComparison.create(filesize1, filesize2);
}
