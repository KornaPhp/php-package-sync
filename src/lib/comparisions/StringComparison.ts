/* eslint-disable no-unused-vars */

const { compareTwoStrings } = require('string-similarity');

export class StringComparison {
    protected _similarityScore: number;

    constructor(protected _string1: string, protected _string2: string) {
        this._similarityScore = compareTwoStrings(this.string1, this.string2);
    }

    static create(string1: string, string2: string) {
        return new StringComparison(string1, string2);
    }

    get similarityScore() {
        return this._similarityScore;
    }

    get string1() {
        return this._string1;
    }

    get string2() {
        return this._string2;
    }

    public meetsRequirement(minimum: number) {
        return this.similarityScore >= minimum;
    }
}
