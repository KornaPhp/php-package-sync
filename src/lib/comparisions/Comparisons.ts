import { FileSizeComparison } from './FileSizeComparison';
import { StringComparison } from './StringComparison';

export class Comparisons {
    public static filesizes(sourceSize: number | undefined, targetSize: number | undefined) {
        return FileSizeComparison.create(sourceSize ?? 0, targetSize ?? 0);
    }

    public static strings(string1: string | undefined, string2: string | undefined) {
        return StringComparison.create(string1 || '', string2 || '');
    }
}
