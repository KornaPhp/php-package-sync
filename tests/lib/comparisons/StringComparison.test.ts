/* eslint-disable no-undef */

import { StringComparison } from '../../../src/lib/comparisions/StringComparison';

it('compares two strings', () => {
    expect(StringComparison.create('one two three', 'one two five'))
        .toMatchSnapshot();
    expect(StringComparison.create('one two three', 'one two three'))
        .toMatchSnapshot();
});

it('compares two equal strings', () => {
    expect(StringComparison.create('one two three', 'one two three').similarityScore)
        .toBe(1.0);
});

it('compares two strings and checks the minimum requirement', () => {
    expect(StringComparison.create('one two three', 'one two three')
        .meetsRequirement(0.3))
        .toBeTruthy();
    expect(StringComparison.create('one two three', 'five six seven')
        .meetsRequirement(0.9))
        .toBeFalsy();
});

it('returns the correct values for stringN properties', () => {
    expect(StringComparison.create('one two three', 'one two four').string1)
        .toBe('one two three');
    expect(StringComparison.create('one two three', 'one two four').string2)
        .toBe('one two four');
});
