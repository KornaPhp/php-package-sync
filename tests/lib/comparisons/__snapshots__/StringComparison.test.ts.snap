// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compares two strings 1`] = `
StringComparison {
  "_similarityScore": 0.5263157894736842,
  "_string1": "one two three",
  "_string2": "one two five",
}
`;

exports[`compares two strings 2`] = `
StringComparison {
  "_similarityScore": 1,
  "_string1": "one two three",
  "_string2": "one two three",
}
`;
